xcm/xcm-executor/src/traits/weight.rs

```diff
@@ -76,11 +76,27 @@ pub trait WeightTrader: Sized {
 	}
 }
 
-impl WeightTrader for () {
+#[impl_trait_for_tuples::impl_for_tuples(30)]
+impl WeightTrader for Tuple {
 	fn new() -> Self {
-		()
+		for_tuples!( ( #( Tuple::new() ),* ) )
 	}
-	fn buy_weight(&mut self, _: Weight, _: Assets) -> Result<Assets, Error> {
-		Err(Error::Unimplemented)
+
+	fn buy_weight(&mut self, weight: Weight, payment: Assets) -> Result<Assets, Error> {
+		for_tuples!( #(
+			if let Ok(assets) = Tuple.buy_weight(weight, payment.clone()) {
+				return Ok(assets);
+			}
```

Comment:
This swallows any errors that might happen along the way. Not sure we want to do that.

Reply:
I think the only thing we care about is if all of them fail, which is the final error below. Otherwise, we simply want to check with each trader if it is possible to buy weight.

Reply:
I think that retry behavior is generally fine, but I think it would be good to at least surface the last error (returned by the last impl of buy_weight).
We should also add some logging (similar to the other XCM tracing I added) and maybe that can be enough.

Reply:
I just want to make sure it's possible to debug this at all (which I fear the current version is not).

Reply:
is a log trace message enough?

Reply:
is a trace log enough?

Reply:
What do you think of:
```Rust
fn buy_weight(&mut self, weight: Weight, payment: Assets) -> Result<Assets, Error> {
		let mut last_error = None;
		for_tuples!( #(
			match Tuple.buy_weight(weight, payment.clone()) {
				Ok(assets) => return Ok(assets),
				Err(e) => { last_error = Some(e) }
			}
		)* );
		log::trace!(target: "xcm::buy_weight", "last_error: {:?}", last_error);
		Err(last_error)
	}
```

Reply:
It's helpful to add a trace log for debugging. Not sure about returning the last error, as it doesn't represent the reason why all traders failed.